tensorboard/notebook.py

```diff
@@ -395,11 +395,15 @@ def _display_colab(port, height, display_handle):
 
 def _display_ipython(port, height, display_handle):
   import IPython.display
-  iframe = IPython.display.IFrame(
-      src="http://localhost:%d" % port,
-      height=height,
-      width="100%",
-  )
+  shell = """
+  <iframe
+    src="javascript:window.location='http://' + parent.location.hostname + ':%PORT%/';"
```

Comment:
This use of a JavaScript URI for the `src` attribute is fairly
surprising/confusing, and also generates a flash of bad content as the
JS expression evaluates to a string. Can we instead explicitly set the
location, using a shell like the following?

```html
<iframe id="tensorboard-frame" …></iframe>
<script>
(function() {
  const frame = document.getElementById("tensorboard-frame");
  const url = new URL("/", window.location);
  url.port = %PORT%;
  frame.contentWindow.location.replace(url);
})();
</script>
```

We also need to preserve the protocol, which may be `https` rather than
`http`; the above shell does that.

If you could verify that something like this works for your use case,
that’d be great.

Reply:
I didn't preserve the protocol because I didn't think the TB port could handle that, but I suppose it's true enough that someone could have enabled this.
I did notice the "flash of bad content" issue you mentioned; the reason I didn't do it quite like you said is that it seemed possible that we could have multiple iframes of the same ID, and I was in a bit too much of a rush at the time to generate unique iframe IDs for each.  Thoughts on that?

Reply:
(Regarding protocol, I note that the original implementation always specified `http://localhost/...`, regardless of whether the enclosing Jupyter notebook would have had been loaded via http or https etc., so I followed the assumption of the original code in that aspect.)

Reply:
Ah, you’re right about the IDs, given that Jupyter output cells aren’t themselves framed (like Colab output cells are).

My first thought was to try something like the following—

```html
<script>
(function() {
  const frame = document.createElement("iframe");
  const url = new URL("/", window.location);
  url.port = %PORT%;
  frame.src = url;
  const me = document.currentScript;
  me.parentNode.insertBefore(frame, me);
})();
</script>
```

—but Jupyter rehydrates scripts by appending them to `document.head`, so
`document.currentScript` actually refers to the wrong element here, and
this doesn’t work.

Perhaps we could fall back to a simple nonce-based solution? Something
like:

```python
import cgi
import json
import random

frame_id = "tensorboard-frame-{:08x}".format(random.randrange(0, 1 << 64))
shell = """
<iframe id="%HTML_ID%" width="100%" height="%HEIGHT%"></iframe>
<script>
(function() {
  const frame = document.getElementById(%JSON_ID%);
  const url = new URL("/", window.location);
  url.port = %PORT%;
  frame.src = url;
})()
</script>
"""
replacements = [
    ("%HTML_ID%", cgi.escape(frame_id, quote=True)),
    ("%JSON_ID%", json.dumps(frame_id)),
    ("%PORT%", "%d" % port),
    ("%HEIGHT%", "%d" % height),
]
for (k, v) in replacements:
  shell = shell.replace(k, v)
IPython.display.HTML(shell)
```

Reply:
That seems reasonable; I'll give it a try.

What about the protocol preservation question?

Reply:
Preserving the protocol wasn’t super important in the old code because
the hostname was hard-coded, too. If preserving the protocol is
difficult for some reason, then it’s not a huge deal to assume HTTP; if
it’s basically free (as with the above shell), then let’s do that.

Reply:
OK.  As long as we're in agreement that the "https" that would be supported here would be the Werkzeug "rudimentary" variety discussed in https://github.com/tensorflow/tensorboard/issues/260 , not a case with nginx or similar, unless somebody works out a way to give TensorBoard one `--host` (on a given port) and nginx terminates SSL through to the same port on a different IP address (i.e., the one visible to the client).

Reply:
> you’re right about the IDs, given that Jupyter output cells aren’t themselves framed (like Colab output cells are).

A maybe simpler alternative I just realized - there's nothing stopping us from having nested IFrames inline here.  I'm trying out that approach now.

Reply:
That didn't pan out - even though the nested javascript works, the nested iframe doesn't work inline without more gymnastics than what I was trying to eliminate in the first place.  :)  Following your suggestion and updating the PR.